### PR TITLE
DataDog-friendly OpenTelemetry metadata

### PR DESCRIPTION
**Pull request by LeartS**
Having opentelemetry metadata like the trace_id and span_id in logs is extremely useful to correlate logs of the same trace, and if the metadata is in the format expected by DataDog, you'll gain some very powerful Logs-APM datadog features.

Currently, we made this work (more or less) by explicitly setting Logger metadata in our instrumenting libraries when spans are created, for example:
* [Teleplug](https://github.com/primait/teleplug/blob/3245cec326c918d69f6d33161cd875eae326f150/lib/teleplug.ex#L36)
* [opentelemetry_absinthe](https://github.com/primait/opentelemetry_absinthe/blob/694b2bd026e8bdfd3961d7c12a0bf545f61cd366/lib/instrumentation.ex#L73)

The problem with this is that each instrumenting library needs to do so, and consistely (internally and with other instrumenting libraries). Additionally, if in your application code you start an internal span (in the same process) using the OpenTelemetry API directly, you must also always remember to do so, otherwise the logs emitted during the internal span will still reference the parent span.

The OpenTelemetry API/SDK, since 1.1.0, automatically adds opentelemetry-related log metadata to processes that have a trace context set.

This is extremely useful as it allows all opentelemetry-instrumented code to have opentelemetry log metadata automatically, without the need for each instrumenting library to set the metadata explicitly whenever setting the current span.

Unfortunately, the metadata set by the SDK doesn't play nicely with DataDog: it doesn't use the keys that datadog expects to make logs-APM correlation work, and additionally it's set a charlists values which are serialized to unsightly list of integers by prima_ex_logger.

This PR introduces a new option to transform the OpenTelemetry metadata to some different representations that work better with Prima's infrastructure.

The new default is the :datadog format, which will transform the SDK opentelemetry metadata into datadog-friendly equivalents. Check the README for other options.

---------------

~This PR should be followed up by other PRs in all of our instrumenting libraries to remove the bespoke setting of logger metadata.~
Done, see linked PRs

---------------

I initially tried to soft-depend and use the OpenTelemetry API to get the span context directly, without relying on the metadata set by the OpenTelemetry SDK, but quickly realized that this can't work because the logger backend is in its own process. I could do some "black magic" to get the opentelemetry context out of the process dictionary using `Process.info`, given that I get the `pid` of the log-emitting process in the metadata. I tried that approach and seems to work, but it felt even more brittle than this solution (= relying on Logger metadata set by the OpenTelemetry SDK).
The advantage of this solution is that it will also take care of removing the recent unsightly metadata that started appearing since opentelemetry api 1.1.0 (see screenshot)
image

### CODE REVIEW

**Review by chess4ever — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -31,12 +31,34 @@ config :logger, :prima_logger,
...
 - **encoder** (module): JSON encoder, default Jason. Tested with Jason, Poison and JSX
+
 - **type** (string): app name
+
 - **environment** (atom): current environment
-- **metadata** (list): custom metadata to append on every log, default []
-- **metadata_serializers** (list): custom serializers for structs found in metadata, default []
+
+- **metadata** (list): custom metadata to append to every log, default `[]`.
+  Note that this has a different meaning that Logger's `metadata` option, which is used to indicate what metadata keys to keep instead!
```
Comment (cpiemontese) [inline README.md:42]:
```suggestion
  Note that this has a different meaning than Logger's `metadata` option, which is used to indicate what metadata keys to keep instead!
```

File: README.md [hunk 2]
```diff
@@ -31,12 +31,34 @@ config :logger, :prima_logger,
...
+
+  The value indicates the "format" of the opentelemetry metadata to use:
+    - `:raw`: leaves the opentelemetry metadata generated by the opentelemetry sdk intact and doesn't change it any way.
+      Not recommended because these are serialized to quite useless list of integers.
+      All other options will remove the opentelemetry SDK metadata from the logs, but will use it to compute the equivalents in other formats.
+    - `:datadog`: sets `dd.trace_id` and `dd.span_id` as per the [DataDog doc on connecting OpenTelemetry Traces and Logs](https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/opentelemetry).
+    This is the default and will allow you to use the extremely useful log / APM correlation features of DataDog.
+    - `:opentelemetry`: sets `otel.trace_id`, `otel.span_id` and `otel.trace_flags'` using the OpenTelemetry hex-encoded TraceId/SpanId formats.
+    - `:detailed`: sets both of the above
+    - `none`: No automatic opentelemetry-related metadata at all in the produced logs.
```
Comment (cpiemontese) [inline README.md:54]:
```suggestion
    - `:none`: No automatic opentelemetry-related metadata at all in the produced logs. 
```

File: lib/prima_ex_logger.ex [hunk 3]
```diff
@@ -75,38 +85,123 @@ defmodule PrimaExLogger do
...
+
+      format when format in [:datadog, :opentelemetry_clean, :detailed] ->
+        otel_metadata =
+          try do
+            metadata
+            |> Keyword.take(@opentelemetry_sdk_metadata_keys)
+            |> opentelemetry_metadata(format)
+          rescue
+            # If the otel metadata is not in the format we expect
+            # (for example because the application may be oerwriting it with arbitrary values)
```
Comment (cpiemontese) [inline lib/prima_ex_logger.ex:133]:
```suggestion
            # (for example because the application may be overwriting it with arbitrary values)
```

**Review by cpiemontese — COMMENTED** [review]
Minor typos

File: README.md [hunk 4]
```diff
@@ -31,12 +31,34 @@ config :logger, :prima_logger,
...
+    This is the default and will allow you to use the extremely useful log / APM correlation features of DataDog.
+    - `:opentelemetry`: sets `otel.trace_id`, `otel.span_id` and `otel.trace_flags'` using the OpenTelemetry hex-encoded TraceId/SpanId formats.
+    - `:detailed`: sets both of the above
+    - `none`: No automatic opentelemetry-related metadata at all in the produced logs. 
+
+  Note that, for this functionality to work, your project must depend on `:opentelemetry_api` >= 1.1 and have functioning opentelemetry instrumentation,
+  for example by using [prima_opentelemetry_ex](https://github.com/primait/prima_opentelemetry_ex).
+
+- **metadata_serializers** (list): custom serializers for structs found in metadata, default `[]`
     - example: `[{Decimal, to_string}]`, will invoke `Decimal.to_string/1` when a `Decimal` struct is found among metadata
```
Comment (tshakah) [inline README.md:60]:
```suggestion
    - example: `[{Decimal, to_string}]`, will invoke `Decimal.to_string/1` when a `Decimal` struct is found in the metadata
```

File: README.md [hunk 5]
```diff
@@ -31,12 +31,34 @@ config :logger, :prima_logger,
...
     - example: `[{Decimal, &Kernel.to_string/1}]`, will invoke `Kernel.to_string/1` when a `Decimal` struct is found among metadata
```
Comment (tshakah) [inline README.md:61]:
```suggestion
    - example: `[{Decimal, &Kernel.to_string/1}]`, will invoke `Kernel.to_string/1` when a `Decimal` struct is found in the metadata
```

File: lib/prima_ex_logger.ex [hunk 6]
```diff
@@ -75,38 +85,123 @@ defmodule PrimaExLogger do
...
+  def alter_opentelemetry_metadata(metadata, settings) do
+    no_otel_metadata = Keyword.drop(metadata, @opentelemetry_sdk_metadata_keys)
+
+    case settings.opentelemetry_metadata do
+      # No opentelemetry-related metadata at all
+      :none ->
+        no_otel_metadata
+
+      # Leave the standard log metadata set by the opentelemetry SDK untouched.
+      # Not recommended when sending logs to datadog, as these produce really ugly metadata
```
Comment (tshakah) [inline lib/prima_ex_logger.ex:121]:
```suggestion
      # Not recommended when sending logs to datadog, as this results in really ugly metadata.
```

File: lib/prima_ex_logger.ex [hunk 7]
```diff
@@ -75,38 +85,123 @@ defmodule PrimaExLogger do
...
+        Keyword.merge(no_otel_metadata, otel_metadata)
+    end
+  end
+
+  @spec opentelemetry_metadata(Logger.metadata(), :datadog | :opentelemetry | :detailed) ::
+          Keyword.t()
+  def opentelemetry_metadata(raw_otel_metadata, format)
+
+  def opentelemetry_metadata([], _format) do
+    # if here, it means the log event did not have the standard opentelemetry sdk metadata
```
Comment (tshakah) [inline lib/prima_ex_logger.ex:147]:
```suggestion
    # If here, it means the log event did not have the standard opentelemetry sdk metadata
```

File: lib/prima_ex_logger.ex [hunk 8]
```diff
@@ -75,38 +85,123 @@ defmodule PrimaExLogger do
...
+    # (incorrect or missing opentelemetry instrumentation), or the application is using
+    # a version of opentelemetry API/SDK older than 1.1, which is when logger metadata was added.
+    []
+  end
+
+  # Opentelemetry metadata in "DataDog format", see:
+  # https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/opentelemetry
+  # With these, DataDog Logs-APM correlation features will work
+  def opentelemetry_metadata(metadata, :datadog) do
+    # Convert 128 bit OpenTelemetry trace id to 64 bit DataDog trace id (take the last 64 bits)
```
Comment (tshakah) [inline lib/prima_ex_logger.ex:158]:
```suggestion
    # Convert 128 bit OpenTelemetry trace ID to 64 bit DataDog trace ID (take the last 64 bits)
```

File: lib/prima_ex_logger.ex [hunk 9]
```diff
@@ -75,38 +85,123 @@ defmodule PrimaExLogger do
...
+      |> Integer.parse(16)
+      |> case do
+        {span_id, ""} -> Integer.to_string(span_id, 10)
+        _ -> ""
+      end
+
+    [dd: [trace_id: dd_trace_id, span_id: dd_span_id]]
   end
 
+  # Opentelemetry metadata in opentelemetry format: 128 bits trace ids and 64 bits span ids, hex-encoded as binaries.
```
Comment (tshakah) [inline lib/prima_ex_logger.ex:186]:
```suggestion
  # Opentelemetry metadata in opentelemetry format: 128 bit trace IDs and 64 bit span IDs, hex-encoded as binaries.
```

File: test/prima_ex_logger_test.exs [hunk 10]
```diff
@@ -139,6 +139,69 @@ defmodule PrimaExLoggerTest do
...
+      assert "otel" not in Map.keys(event["metadata"])
+    end
+
+    test "values are computed correctly based on the raw opentelemetry metadata" do
+      io =
+        capture_io(fn ->
+          logger = new_logger(opentelemetry_metadata: :detailed)
+
+          log(logger, "hello world!", :info,
+            # this is the log metadata automatically added by opentelemetry sdk >= 1.1.0
```
Comment (tshakah) [inline test/prima_ex_logger_test.exs:185]:
```suggestion
            # This is the log metadata automatically added by opentelemetry sdk >= 1.1.0
```

**Review by tshakah — COMMENTED** [review]
Minor documentation suggestions

**Comment by LeartS** [comment]
Rebased on latest master and version bump.

Marked as a "major" upgrade because, by default, these new changes will automatically produce new metadata and automatically remove some metadata that was in the original log event.

**Review by michelangelomo — DISMISSED** [review]

**Review by cpiemontese — APPROVED** [review]
I'd say it looks good and we can merge it!

**Pull request merged**